update_cpes.py

```diff
@@ -0,0 +1,201 @@
+
+def update_cpes(xml_file, cpe_vp_map, r7_vp_map):
+    parser = etree.XMLParser(remove_comments=False)
+    doc = etree.parse(xml_file, parser)
+
+    for fingerprint in doc.xpath('//fingerprint'):
+
+        # collect all the params, grouping by os and service params that could be used to compute a CPE
+        params = {}
+        for param in fingerprint.xpath('./param'):
+            name = param.attrib['name']
+            # remove any existing CPE params
+            if re.match(r'^.*\.cpe\d{0,2}$', name):
+                param.getparent().remove(param)
+                continue
+
+            match = re.search(r'^(?P<fp_type>hw|os|service(?:\.component)?)\.', name)
+            if match:
+                fp_type = match.group('fp_type')
+                if not fp_type in params:
+                    params[fp_type] = {}
+                if name in params[fp_type]:
+                    raise ValueError('Duplicated fingerprint named {} in {}'.format(name, fingerprint.attrib['pattern']))
+                params[fp_type][name] = param
+
+
+        # for each of the applicable os/service param groups, build a CPE
+        for fp_type in params:
+            if fp_type == 'os':
+                cpe_type = 'o'
+            elif fp_type.startswith('service'):
+                cpe_type = 'a'
+            elif fp_type == 'hw':
+                cpe_type = 'h'
+            else:
+                raise ValueError('Unhandled param type {}'.format(fp_type))
+
+            # extract the vendor/product/version values from each os/service group,
+            # using the static value ('Apache', for example) when pos is 0, and
+            # otherwise use a value that contains interpolation markers such that
+            # products/projects that use recog content can insert the value
+            # extracted from the banner/other data via regex capturing groups
+            fp_data = {
+                'vendor': None,
+                'product': None,
+                'version': '-',
+            }
+            for fp_datum in fp_data:
+                fp_datum_param_name = "{}.{}".format(fp_type, fp_datum)
+                if fp_datum_param_name in params[fp_type]:
+                    fp_datum_e = params[fp_type][fp_datum_param_name]
+                    if fp_datum_e.attrib['pos'] == '0':
+                        fp_data[fp_datum] = fp_datum_e.attrib['value']
+                    else:
+                        fp_data[fp_datum] = "{{{}}}".format(fp_datum_e.attrib['name'])
+
+            vendor = fp_data['vendor']
+            product = fp_data['product']
+            version = fp_data['version']
+
+            # build a reasonable looking CPE value from the vendor/product/version,
+            # lowercasing, replacing whitespace with _, and more
+            if vendor and product:
+                if not cpe_type in cpe_vp_map:
+                    print(cpe_type)
+                    logging.error("Didn't find CPE type %s", cpe_type)
```

Comment:
Adding the vendor/product to this output would provide useful context.
For example:
#### Existing:
```
a
ERROR:root:Didn't find CPE type a
```
#### Possible output 
using `logging.error("Didn't find CPE type '%s' for '%s %s'", cpe_type, vendor, product)`
```
a
ERROR:root:Didn't find CPE type a for Oracle XML DB
```